src/mds/MDCache.cc

```diff
@@ -8515,8 +8515,11 @@ int MDCache::path_traverse(const MDRequestRef& mdr, MDSContextFactory& cf,
 	  lov.add_xlock(&dn->lock);
 	} else {
 	  // force client to flush async dir operation if necessary
-	  if (cur->filelock.is_cached())
+	  if (cur->filelock.is_cached() &&
+	      !(mdr->lock_cache &&
+		static_cast<const MutationImpl*>(mdr->lock_cache)->is_wrlocked(&cur->filelock))) {
```

Comment:
Deciding if we need to take a lock by checking if it's already taken is a huge code smell. How is this safe? If it is safe, please comment with some justification.

Is there really not a way to know up front what the expected and needed state is? How do we ensure that the lock state doesn't change, if we don't own the lock?

Reply:
Once a `lock_cache` is created the locks belong to it won't be dropped until the `lock_cache` is removed.

And for a `lock_cache` it will hold a reference for each `mdr`. So here it will be safe that to check whether the `cur->filelock` is owned by the `lock_cache`'s or not.